### PR TITLE
Type erased prototype

### PR DESCRIPTION
**Pull request by SimonHeybrock**

### CODE REVIEW

File: src/type-erased-prototype/test/dataset_iterator_test.cpp [hunk 1]
```diff
@@ -275,3 +279,55 @@ TEST(GetterMixins, compilation_test) {
   EXPECT_EQ(view.getDouble(), 1.5);
   EXPECT_EQ(view.getInt(), 1);
 }
+
+template <class T> constexpr int type_to_id();
+template <> constexpr int type_to_id<double>() { return 0; }
+template <> constexpr int type_to_id<int>() { return 1; }
+template <> constexpr int type_to_id<char>() { return 2; }
```
Comment (OwenArnold) [inline src/type-erased-prototype/test/dataset_iterator_test.cpp]: For every `T` an explicit specialisation is required, and the author must ensure that the corresponding `int` `id` is unique. Maybe something like https://www.boost.org/doc/libs/1_46_0/libs/preprocessor/doc/ref/counter.html would help?

File: doc/type-erased-prototype.md [hunk 2]
```diff
@@ -163,6 +163,14 @@ Furthermore, `Dataset` will cover many other cases that are currently impossible
   - Changes/complicates the way of interaction with other libraries such as `numpy` and `gsl`.
     - Vanilla `numpy` does not appear to have good support for arrays of value/error tuples, i.e., this does not appear to be a good solution.
   - Would make vectorization more difficult.
+  - Given the disadvantages of storing values and errors as `DataPoint`, we have the option to keep value and error separate in the `Dataset`, but can nevertheless provide access via `DataPoint` for convenience:
+    ```cpp
+    DatasetView<Variable::DataPoint> view(dataset);
```
Comment (OwenArnold) [inline doc/type-erased-prototype.md]: In this example at least, I don't see how we would be able to deal with multiple `DataPoint` in the same `Dataset`.. Maybe I'm missing something?

Comment (SimonHeybrock) [reply doc/type-erased-prototype.md]: If the `Variable` tag is ambiguous you need to use the variable name to access it,
```cpp
DatasetView<Variable::DataPoint> view(dataset, "sample42");
```

File: doc/type-erased-prototype.md [hunk 3]
```diff
@@ -0,0 +1,400 @@
...
+  - Support variables with non-constant size in a certain dimension, e.g., bin edges and counts that have a different length in the time-of-flight dimension for every point in the spectrum dimension.
+    Without having tried, this should be possible without too much trouble (provided that we do not require to *change* the individual lengths) and would only prevent certain operations, such as selecting a time-of-flight slice.
+    - Need a (hidden?) index variable that provides the start of data for each new histogram.
+      Can we simply use the `Histogram` variable, given that it stores offsets to the underlying data anyway?
+
+- For more convenient handling of `Dataset`, it might be beneficial to combine value and error into `struct DataPoint { double value; double error; };`.
+  This avoid difficulties related to handling two separate variables for values and errors, in particular the need to always access two variables at the same time and the need for a mechanism linking a specific value variable to its error variable.
+  There are a couple of non-negligible disadvantages:
+  - Changes/complicates the way of interaction with other libraries such as `numpy` and `gsl`.
```
Comment (martyngigg) [inline doc/type-erased-prototype.md]: I think this would then basically require copies all over the place to interface with this libraries? I agree that it makes this not really viable.

File: doc/type-erased-prototype.md [hunk 4]
```diff
@@ -0,0 +1,400 @@
...
+    - Vanilla `numpy` does not appear to have good support for arrays of value/error tuples, i.e., this does not appear to be a good solution.
+  - Would make vectorization more difficult.
+  - Given the disadvantages of storing values and errors as `DataPoint`, we have the option to keep value and error separate in the `Dataset`, but can nevertheless provide access via `DataPoint` for convenience:
+    ```cpp
+    DatasetView<Variable::DataPoint> view(dataset);
```
Comment (martyngigg) [inline doc/type-erased-prototype.md]: I like this idea as at least it forces clients to think that a value does have an associated error.

File: doc/type-erased-prototype.md [hunk 5]
```diff
@@ -0,0 +1,400 @@
...
+
+- Variables that are used as an axis to index into the `Dataset` should *not* be verified on write.
+  This has been implemented for ensuring unique spectrum numbers as part of `IndexInfo`, which was introduced as part of refactoring for supporting scanning instruments and MPI.
+  The result was code that is probably hard to maintain, in particular since it required changes outside `IndexInfo`, in particular `ExperimentInfo`, `MatrixWorkspace`, and `ISpectrum`.
+  - Instead, do verification lazily only when an axis is actually used for accessing the `Dataset`.
+  - Results in some overhead (and MPI communication), but does not appear to happen too frequently and should not be an issue for performance.
+
+- A dimension in the `Dataset` is effectively the same as a unit.
+  For example,
+  - `Dimension::Tof` could be used to imply that the unit is time-of-flight microseconds, converting the unit could be done by changing the dimension label to, e.g., `Dimension::dSpacing`.
```
Comment (martyngigg) [inline doc/type-erased-prototype.md]: What should something like `ScaleX` do, i.e. I want my tof in milliseconds for example? Should we have a `Dimension::Custom` or `Dimension::User`?

File: doc/type-erased-prototype.md [hunk 6]
```diff
@@ -0,0 +1,400 @@
...
+  - This is the same issue that `HistogramData::Histogram` attempted but failed to solve:
+    At a high level `Histogram` knows that it stores standard-deviations and the API can return variances if required.
+    However, low level access is sometimes used to square the internal error values, such that there is a mismatch between the actual content and the "guarentee" that `Histogram` gives.
+    Without storing compile-time quantities this is a problem that cannot be solved, so we will have to live with it.
+
+  There are a couple of options for having units attached to a variable in a `Dataset`:
+  - A variable could be a quantity, basically something like `std::pair<UnitTag, std::vector<double>>`.
+    - Not all variables would need to have a unit, but common ones such as Y and E would.
+  - The unit tag could be a separate variable:
+    - Any operation on `Dataset` that wants to handle units could simply handle it separately?
```
Comment (martyngigg) [inline doc/type-erased-prototype.md]: I think this could be nice as again a lot of operations don't really care about the unit.

File: doc/type-erased-prototype.md [hunk 7]
```diff
@@ -0,0 +1,400 @@
...
+    - Enforce unit variable to be zero-dimensional?
+
+  `Variable::Histogram` should link to the unit such that operations with histogram are safe.
+  - The unit link would then handle things like preventing adding a time-of-flight histogram to a d-spacing histogram).
+  - Operations with histograms will have only const access to the unit (since there are many histograms but just one unit in the `Dataset`), preventing in-place operation with histograms if the unit changes.
```
Comment (martyngigg) [inline doc/type-erased-prototype.md]: I think it would be a shame to lose in-place unit conversion for at least something like convert units. A common operation for some of the big diffraction machines here is to load and immediately convert to dspacing and having to allocate the new 30Gb dataset to just delete the old one seems harsh.

File: doc/type-erased-prototype.md [hunk 8]
```diff
@@ -0,0 +1,400 @@
...
+- `Dimension::ScanIndex`, for scanning instrument (sync scan only), applies to `Position` and `Rotation` variables (currently part of `DetectorInfo`) and `ScanIntervals`.
+- `Dimension::Spectrum`, applies to all variables that depend on the spectrum (but not the detector). Key examples are `SpectrumNumbers`, `SpectrumDefinitions` (mapping from spectrum to one or multiple detectors), `SpectrumInfo` (or an equivalent representation based in individual independent variables for positions, rotations, ...), `EventLists`, and histogrammed data.
+- `Dimension::Tof`, applies mainly to variables X, Y, and E of histogrammed data, but see also discussion above regarding variable-length histograms (`Tof` is probably a bad name since it may be anything derived from it, but see discussion on units above).
+- Multiple masking could be supported, e.g., for both `Dimension::Detector` to mask detectors as well as for `Dimension::Spectrum` and `Dimension::Tof` for masking (bins of) spectra.
+  - Should we have first-class masking support, i.e., as a feature built into `Dataset` that is handled automatically in many cases?
```
Comment (martyngigg) [inline doc/type-erased-prototype.md]: I'm not clear what advantages it offers if we don't handle masking automatically.

File: doc/type-erased-prototype.md [hunk 9]
```diff
@@ -0,0 +1,400 @@
...
+// e.g., instrument, detector info, spectrum info, and mapping to detectors.
+
+// Dimensions can be added at any time.
+// Add axis with new dimension
+d.add<Variable::Polarization>({Spin::Up, Spin::Down}, Dimension::Polarization, "pol");
+// Extend the histogram variable into the new dimension (data zero initialized?).
+d.extend(Variable::Histogram, Dimension::Polarization);
+
+// Copy constructor. All variables are automatically shared.
+auto d_copy(d);
```
Comment (martyngigg) [inline doc/type-erased-prototype.md]: This would be a shallow copy, yes?

File: doc/type-erased-prototype.md [hunk 10]
```diff
@@ -0,0 +1,400 @@
...
+## Interface mock-up
+
+Naming is just for illustration, may be very different in the actual implementation.
+
+```cpp
+Dataset d;
+
+// Imagine arbitrary initialization here
+
+// Variables can be added (or removed) at any time.
```
Comment (martyngigg) [inline doc/type-erased-prototype.md]: For most cases we would want some type of "standard" set of variables - would we provide some factory methods for those cases?

File: doc/type-erased-prototype.md [hunk 11]
```diff
@@ -0,0 +1,349 @@
...
+  - At runtime or compile-time?
+  - Do we need things like expression templates to make this work?
+    Does not play nicely with `auto`.
+  - Is it worth the effort, given that we may often be limited by I/O?
+
+## Example
+
+A typical `Dataset`, equivalent to one of our current workspaces could contain the following variables and dimensions:
+
+- Experiment logs and sample with (typically) no attached dimension.
```
[the remark this answers is not in the document]
Comment (martyngigg) [reply doc/type-erased-prototype.md]: What about possible units for log data?

File: doc/type-erased-prototype.md [hunk 12]
```diff
@@ -0,0 +1,400 @@
...
+    template <class Tag>
+    variable_type_t<Tag> &get() conditional_const;
+  };
+  ```
+  Additionally access via a name will probably be provided.
+
+- We will need to support "axis" variables that are bin edges, i.e., are by 1 longer than the size of the dimension.
+  If we iterate over all dimensions, what should the iteration for bin edges do?
+  - Not possible?
+  - Support iterating "bins", a virtual variable providing center, left edge, right edge, and width?
```
[the remark this answers is not in the document]
Comment (SimonHeybrock) [reply doc/type-erased-prototype.md]: Iterating only over the bin edges on their own is of course still possible, e.g., by directly accessing the respective variable.

If you are refering to iterating over bin edges jointly with other variables, I am not sure what type of access you have in mind for the variables that are not defined at the bin edges but in the bins?

File: doc/type-erased-prototype.md (hunk 3 above)
Comment (SimonHeybrock) [reply doc/type-erased-prototype.md]: Not necessarily. `gsl` fitting routines support a spread for y and e (which would be 2 in this case).

For `numpy`, the `dtype` might be changed to pairs (resulting in arrays where the default math operations don't do the right thing). There is also https://pythonhosted.org/uncertainties/numpy_guide.html. Both solution are much less wide spread and thus not a good solution.

That being said, I nevertheless agree with your conclusion that is is not really viable.

File: doc/type-erased-prototype.md (hunk 5 above)
Comment (SimonHeybrock) [reply doc/type-erased-prototype.md]: Yes, there are probably a couple of solutions:

- Support a scale factor for the dimension (I believe `boost::units` has something similar).
- Support a couple of standard cases, e.g., `Dimension::Tof::Microsecond` and `Dimension::Tof::Millisecond`.
- Something like what you say.

File: doc/type-erased-prototype.md (hunk 6 above)
Comment (SimonHeybrock) [reply doc/type-erased-prototype.md]: I am currently favoring having the unit on the variable. For standard operations that we provide such as addition, multiplication, ... . this ensures that the unit is always updated, even when working with individual variables instead of `Dataset`. For lower-level operations implemented on items of variables the unit would not be visible anyway, so operations that do not care would not see it.

Basically, I you implement an operation for a whole variable (containing many items), you should probably care about the unit. This case might be rare.

If you work on a lower level, you will not see it (except for `Histogram`, but again if your operation is in-place or just takes a copy of the input histogram you get the right (or unchanged) unit automatically).

File: doc/type-erased-prototype.md (hunk 7 above)
Comment (SimonHeybrock) [reply doc/type-erased-prototype.md]: I do not think we would lose that. It would be no big trouble to implement `ConvertUnits` in-place. What I am saying is that using the default "convenience" interface via histograms you cannot change the unit. But we can simply bypass that for specific cases that require better performance.

File: doc/type-erased-prototype.md (hunk 8 above)
Comment (SimonHeybrock) [reply doc/type-erased-prototype.md]: You are right that there should definitely automatical handling by standard operations where possible. What is not so clear to me is how we can reasonably and unambiguously define this in all cases.

File: doc/type-erased-prototype.md (hunk 10 above)
Comment (SimonHeybrock) [reply doc/type-erased-prototype.md]: Yes, on the one hand complete factory methods for "standard" datasets. Additionally there could be helpers for standard components.
```cpp
/// Add all variables for instrument, including those replacing DetectorInfo, SpectrumInfo, and spectrum-detector mapping.
void addInstrument(Dataset &d);
/// Add Variable::Value, Variable::Error, Variable::Histogram
void addHistogram(Dataset &d, const std::string &name);
```

File: doc/type-erased-prototype.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (SimonHeybrock) [reply doc/type-erased-prototype.md]: The current logs would be a single item of a variable, so an overall unit might not make sense. So if we want units we have two options:
1. Support in within our current log data structure.
2. Split our log data structure into variables, such that each variable contains only a single log and we can thus use the "standard" unit mechanism descirbed above.

**Comment by SimonHeybrock** [comment]
Merging this for now as there are no more comments, and to facilitate follow-up work.

**Comment by SimonHeybrock** [comment]
I have tagged this state as [dataset-v1](https://github.com/mantidproject/workspace-sandbox/releases/tag/dataset-v1) for future reference.
